java/google/registry/rde/imports/RdeDomainImportAction.java

```diff
@@ -145,17 +161,70 @@ private RdeImportUtils createRdeImportUtils() {
     public void map(JaxbFragment<XjcRdeDomainElement> fragment) {
       final XjcRdeDomain xjcDomain = fragment.getInstance().getValue();
       try {
-        logger.infofmt("Converting xml for domain %s", xjcDomain.getName());
         // Record number of attempted map operations
         getContext().incrementCounter("domain imports attempted");
+
         logger.infofmt("Saving domain %s", xjcDomain.getName());
         ofy().transact(new VoidWork() {
           @Override
           public void vrun() {
-            DomainResource domain =
-                XjcToDomainResourceConverter.convertDomain(xjcDomain);
-            getImportUtils().importDomain(domain);
+            HistoryEntry historyEntry = createHistoryEntryForDomainImport(xjcDomain);
+            BillingEvent.Recurring autorenewBillingEvent =
+                createAutoRenewBillingEventForDomainImport(xjcDomain, historyEntry);
+            PollMessage.Autorenew autorenewPollMessage =
+                createAutoRenewPollMessageForDomainImport(xjcDomain, historyEntry);
+            DomainResource domain = XjcToDomainResourceConverter.convertDomain(
+                xjcDomain, autorenewBillingEvent, autorenewPollMessage);
             getDnsQueue().addDomainRefreshTask(domain.getFullyQualifiedDomainName());
+            // Keep a list of "extra objects" that need to be saved along with the domain
+            // and add to it if necessary.
+            ImmutableSet<Object> extraEntitiesToSave =
+                getImportUtils().createIndexesForEppResource(domain);
+            // Create speculative server approval entities for pending transfers
+            if (domain.getTransferData().getTransferStatus() == TransferStatus.PENDING) {
+              TransferData transferData = domain.getTransferData();
+              checkArgumentNotNull(transferData,
+                  "Domain %s is in pending transfer but has no transfer data",
+                  domain.getFullyQualifiedDomainName());
+              Money transferCost = getDomainRenewCost(
+                  domain.getFullyQualifiedDomainName(),
+                  transferData.getPendingTransferExpirationTime(),
+                  transferData.getExtendedRegistrationYears());
+              // Create speculative entities in anticipation of an automatic server approval.
+              ImmutableSet<TransferServerApproveEntity> serverApproveEntities =
+                  createTransferServerApproveEntities(
+                      transferData.getPendingTransferExpirationTime(),
+                      domain.getRegistrationExpirationTime()
+                          .plusYears(transferData.getExtendedRegistrationYears()),
```

Comment:
Looks like you found a bug. `transferData.getPendingTransferExpirationTime()` should not be used below, I need to calculate the new expiration using `extendRegistrationWithCap(..)`